packages/blade/src/components/AutoComplete/_decisions/decisions.md

```diff
@@ -0,0 +1,288 @@
+3. Controlled state of selected items
+
+```jsx
+const allCities = ['Mumbai', 'Pune', 'Bangalore']
+const [filteredValues, setFilteredValues] = React.useState(allCities);
+const [selections, setSelections] = React.useState([]);
+
+// ...
+
+<AutoComplete
+  // For controlling the selections
+  value={selections}
+  onChange={({ name, values }) => {
+    setSelections(values)
+  }}
+  // For controlling the input value
+  inputValue={}
+  onInputValueChange={(inputValue) => {
+    // Filter and set the state here
+    const filteredCityValues = allCities
+      .filter((city) => city.includes(inputValue))
+      .map((city) => city.toLowerCase());
+
+    setFilteredValues(filteredCityValues);
+  }}
+  // Control the filtering of items (Make sure to pass array of values here)
+  filteredValue={filteredValues}
+/>
+
+// ...
+
+<ActionList>
+  {allCities.map((city) => <ActionListItem title={city} value={city.toLowerCase()} />)}
+</ActionList>
+```
+
+Complex flows such as [Adding New Item to ActionList](https://www.figma.com/file/jubmQL9Z8V7881ayUD95ps/Blade---Payment-Light?node-id=47091%3A711978&mode=dev) should be handled with Controlled API.
+
+Checkout the [Changelog](https://github.com/razorpay/blade/pull/1521#issue-1842703697) to see alternate API and limitations of it
+
+<details>
+<summary>Show Full Code</summary>
+
+```jsx
+const allCities = [
+  {
+    title: 'Mumbai',
+    value: 'mumbai'
+  },
+  {
+    title: 'Pune',
+    value: 'pune'
+  },
+  {
+    title: 'Bangalore',
+    value: 'bangalore'
+  },
+];
+
+const ControlledAutoComplete = () => {
+  // Controlled Filtering
+  const [filteredValues, setFilteredValues] = React.useState(allCities);
+  // Controlled Input Value
+  const [currentInputValue, setCurrentInputValue] = React.useState('');
+  // Controlled Selections
+  const [currentSelections, setCurrentSelections] = React.useState<string[]>([]);
+
+  return (
+    <Dropdown>
+      <AutoComplete
+        label="Select City"
+        // Controlled input value
+        // Sets the input value of input element in AutoComplete
+        inputValue={currentInputValue}
+        // Triggers when input value changes
+        onInputValueChange={(inputValue) => {
+          if (inputValue) {
+            // filtering logic
+            const allCityValues = allCities.map((city) => city.value);
+            const filteredCities = allCityValues.filter((cityValue) => cityValue.includes(inputValue));
+            setFilteredValues(filteredCities);
+          } else {
+            setFilteredValues(allCityValues);
+          }
+          setCurrentInputValue(inputValue);
+        }}
+        // Selected Items in AutoComplete
+        value={currentSelections}
+        onChange={({ values }) => {
+          setCurrentSelections(values);
+        }}
+        // controls the values that should be shown to user
+        filteredValues={filteredValues}
+      />
+      <DropdownOverlay>
+        <ActionList>
```

Comment:
question: Does ActionList automatically handle the already selected items and move it down to the bottom, because showing it in the dropdown could cause clutter and could hinder the user interaction, right? Or is it upto the dev. I meant something like - `[...selectableItems, ...selectedItems]` ?

Reply:
No it will keep its place because moving items might be a bit more chaotic UX than keeping them there and having check mark next to it.

We have a good visual indicator of checkbox in actionlistitems so it will distinguish it properly with non-selected items. We follow similar UX as [AriaKit](https://ariakit.org/examples/combobox-multiple)